Absolute Geolocation Error for CSLC-S1

vbrancat opened the pull request:
This PR aims at evaluating the Absolute Geolocation Error for OPERA CSLC-S1. 
The script `ale.py` requires as input:
1. A CSLC-S1 product
2. A CSV file containing the surveyed Corner Reflector locations (CR)

For each CR location contained in the CSLC-S1 product, the code:
1. Identifies the peak corresponding to the CR and computes the image coordinates
2. Computes ALE by differencing the computed CR location with the surveyed CR location

Results can be saved in a CSV file on disk or displayed on screen with a plot

seongsujeong commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+    '''
+
+    with h5py.File(cslc_file, 'r') as h5:
+        x_vect = h5['science/SENTINEL1/CSLC/grids/x_coordinates'][()]
+        y_vect = h5['science/SENTINEL1/CSLC/grids/y_coordinates'][()]
```
Please note that there is half pixels of difference in the upper left corner's coordinates when the raster is loaded as `HDF5` and `NETCDF`. Based on what I've seen, the coordinates as NETCDF raster made more sense becuase the number was a multiple of 30.0

seongsujeong commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+    return cslc_poly
+
+
+def get_clsc_epsg(cslc_file):
```
```suggestion
def get_cslc_epsg(cslc_file):
```

seongsujeong commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+    ovs_factor: np.int
+        Oversampling factor
+    margin: int
+        Margin
```
```suggestion
        half of the margin's size  in pixels
```

seongsujeong reviewed (commented):
This is a good generalization of the code I've written [here](https://github.com/seongsujeong/compass_util/blob/main/compass_util.py). My initial comments are as below.

vbrancat commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+    '''
+
+    # Extract an area around x_loc, y_loc
+    img = arr[(y_loc - margin):(y_loc + margin),
```
It is an abbreviation for location. Do you want me to change the name of the variable if you think it is not sufficiently descriptive?

scottstanie commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+        raise FileNotFoundError(err_str)
+
+    # Open corner reflector file
+    cr_df = pd.read_csv(cr_file)
```
`df` is the typical variable name for a pandas `DataFrame`: https://pandas.pydata.org/docs/user_guide/10min.html

vbrancat replied:
`df` stands for `DataFrame`. This is a typical variable abbreviation when using a `pandas` `DataFrame`

seongsujeong replied:
@scottstanie @vbrancat Good to know. Thanks for the information!

vbrancat commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+        cr_lon = row['Longitude (deg)']
+        cr_loc = geometry.Point(cr_lon, cr_lat)
+
+        if cslc_poly.contains(cr_loc):
```
Are you suggesting to apply a margin to the CR location?

seongsujeong replied:
Yes. I think applying the margin in the image coordinate would be ideal. We can attempt to extract an image chip, and check if there are any "no value" in the chip. In that case we drop the image chip and the CR.

vbrancat replied:
This has been fixed in the latest commit.

vbrancat replied:
Not sure I totally agree with it. We are just checking that the CR position from the CSV is contained in the valid portion of the burst. In which case would the margin be useful? Like if the CR is at the boundary of a burst?

seongsujeong replied:
>In which case would the margin be useful? Like if the CR is at the boundary of a burst?

If we don't have enough margin around the peak, then we can lose some signals around the peak in the SLC, which can affect the peak location in the oversampled SLC. In the case in the figure below, the first ovresampled SLC patch has peak at (2031, 2016), and the second one was at (2056, 2025)

![output_20230410_1511](https://user-images.githubusercontent.com/27862199/231008708-550212b7-b492-49cc-ab8b-7258483c5d2c.png)

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,325 @@
+    '''
+    with h5py.File(cslc_file, 'r') as h5:
+        poly = h5['science/SENTINEL1/identification/bounding_polygon'][()]
+    cslc_poly = wkt.loads(poly)
```
you probably need to decode

```suggestion
    cslc_poly = wkt.loads(poly.decode('utf-8'))
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,398 @@
+
+    if plot_ale:
+        fig, ax = plt.subplots(figsize=(8, 6))
+        sc = ax.scatter(cr_df['ALE_Y'], cr_df['ALE_X'], s=200, alpha=0.8,
```
X and Y are swapped?
```suggestion
        sc = ax.scatter(cr_df['ALE_X'], cr_df['ALE_Y'], s=200, alpha=0.8,
```

vbrancat replied:
Resolved, thanks.

seongsujeong reviewed (commented):
More comments are as below
 - Seongsu

vbrancat commented:
@seongsujeong and @hfattahi the latest commits have:
1) Syncronized peak and geolocation error detection with Jeong's algorithm available here: https://github.com/seongsujeong/compass_util/blob/3aa086cdce903ce3853557592100563b69ceb05b/compass_util.py#L122. I have checked that the two codes produce the same geolocation error and they do. I have verified this on multiple bursts at multiple dates. Happy to point you to the data location
2) Expose more options in terms of polarization, margin, oversampling factor etc
3) Try to generalize the code to read NISAR GSLCs

Let me know your feedbacks

vbrancat commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,398 @@
+    cr_df['ALE_X'] = cr_df['CR_X_CSLC_PEAK'] - cr_df['CR_X']
+    cr_df['ALE_Y'] = cr_df['CR_Y_CSLC_PEAK'] - cr_df['CR_Y']
+
+    if csv_output_file is not None:
```
Done.

vbrancat commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,398 @@
+    optional.add_argument('-i', '--plot-ale', dest='plot_ale', default=False,
+                          help='Plot ALE')
+    optional.add_argument('-t', '--set', dest=set, default=False,
+                          help='If True, correct corner reflector positions for Solid Earth Tides')
```
Done.

vbrancat commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,398 @@
+                          help='Save ALE in CSV file (default: None)')
+    optional.add_argument('-i', '--plot-ale', dest='plot_ale', default=False,
+                          help='Plot ALE')
+    optional.add_argument('-t', '--set', dest=set, default=False,
```
Resolved, thanks.

vbrancat commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,398 @@
+
+    # Generate x_vect and y_vect
+    x_vect = geo_trans[0] + np.arange(0, width)*x_spac
+    y_vect = geo_trans[1] + np.arange(0, length)*y_spac
```
Addressed, thanks.

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+import pandas as pd
+import pysolid
+import shapely.wkt as wkt
+from osgeo import gdal
```
```suggestion
from osgeo import gdal, osr
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pysolid
```
I suggest to move this import to `correct_cr_tides` function:
```suggestion

```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+        # the CR from the pandas dataframe
+        if cslc_poly.contains(cr_loc):
+            # Convert corner lat/lon coordinates in UTM
+            cslc_epsg = get_cslc_epsg(cslc_file, mission_id=mission_id)
```
```suggestion
            cslc_epsg = get_cslc_epsg(cslc_file, mission_id=mission_id, pol=pol)
```

vbrancat replied:
Not sure why we need to specify the polarization here. We are reading the `epsg` from the `projection` dataset which should be common for every polarization

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+
+    # Transform CR coordinates to UTM
+    cslc_epsg = get_cslc_epsg(cslc_file,
+                              mission_id=mission_id)
```
```suggestion
                              mission_id=mission_id, pol=pol)
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+    if mission_id == 'S1':
+        cslc_path = '/science/SENTINEL1/CSLC/grids/'
+    elif mission_id == 'NI':
+        cslc_path = '/science/LSAR/GSLC/grids/'
```
Need to pass frequency for NISAR
```suggestion
        cslc_path = '/science/LSAR/GSLC/grids/frequency'
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+    return cslc_poly
+
+
+def get_cslc_epsg(cslc_file, mission_id='S1'):
```
```suggestion
def get_cslc_epsg(cslc_file, mission_id='S1', pol='VV'):
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+    if mission_id == 'S1':
+        epsg_path = '/science/SENTINEL1/CSLC/grids/projection'
+    elif mission_id == 'NI':
+        epsg_path = '/science/LSAR/GSLC/frequencyA/projection'
```
```suggestion
        with h5py.File(cslc_file, 'r') as h5:
+            frequencies=h5["/science/LSAR/identification/listOfFrequencies"][()]
+            freq = frequencies[0].decode('utf-8')
+            frequency = f'frequency{freq}'
+
+        dataset_path = f'NETCDF:{cslc_file}://science/LSAR/GSLC/grids/{frequency}/{pol}'
+        ds = gdal.Open(dataset_path, gdal.GA_ReadOnly)
+        s = osr.SpatialReference(wkt=ds.GetProjection()).ExportToProj4()
+        crs = CRS.from_proj4(s)
+        epsg = crs.to_epsg()
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+        err_str = f'{mission_id} is not a valid mission identifier'
+        raise ValueError(err_str)
+
+    with h5py.File(cslc_file, 'r') as h5:
+        epsg = h5[epsg_path][()]
```
```suggestion
```

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,536 @@
+        Corner reflector position along Y-direction corrected
+        for Solid Earth tide
+    '''
+    # Get geocode SLC sensing start and stop
```
```suggestion
    import pysloid
    # Get geocode SLC sensing start and stop
```

hfattahi reviewed (commented):
Few suggestions to make it work with NISAR GSLC

vbrancat commented:
Thank you, @hfattahi! I addressed your comments. It would be great if I can have a toy GSLC so I can test my changes. the test on OPERA CSLC is pending and needs to be performed for the last set of changes

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,551 @@
+
+
+def correct_cr_tides(cslc_file, cr_lat, cr_lon,
+                     mission_id='S1', pol=pol):
```
```suggestion
                     mission_id='S1', pol='VV'):
```

seongsujeong reviewed (commented):
@vbrancat Thanks for your work at this PR. As I mentioned in the comment above, the result from this PR looks consistent to my code. I've left couple of suggestions in this iteration.

One more suggestion is to add this `ale.py` to `setup.cfg` in case you want this code be installed as a script when installing COMPASS.

hfattahi commented on src/compass/utils/ale.py:
```diff
@@ -0,0 +1,551 @@
+    if mission_id == 'S1':
+        cslc_path = '/science/SENTINEL1/CSLC/grids/'
+    elif mission_id == 'NI':
+        cslc_path = '/science/LSAR/GSLC/grids/frequencyA'
```
```suggestion
        cslc_path = '/science/LSAR/GSLC/grids/frequencyA/'
```

hfattahi reviewed (approved):
Thanks @vbrancat for making the changes to make the script work with NISAR GSLC. I have to fix couple of things on NISAR RSLC and GSLC itself before further testing. But that is beyond this PR and should not hold this PR. The current PR looks good to me. Thank you for your nice work.

seongsujeong commented:
> @seongsujeong not sure why I am not able to comment on your posted comment, do you think that a fixed margin would do the trick? If that is the case, you much would you recommend?

I think fixed margin would be okay as long as we set up enough margins around the peak. In my pessimistic estimate I would say about 6m in geolocation, about 3 pixels in peak radius, (so, 3*sqrt(5*10)=21.2), and 1.5 of safe factor. That would be about 33m, so I would suggest `35m` of margins.

vbrancat commented:
@seongsujeong all your comments have been addressed :D

seongsujeong reviewed (approved):
Thanks for addressing my comments and suggestions. The PR looks good to me. Nice work!
